<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Model nonlinear dynamic eQTL effects in continuous cell states â€¢ scDesignPop</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Model nonlinear dynamic eQTL effects in continuous cell states">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDesignPop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9009</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDesignPop.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/scDesignPop-batch-effect.html">Modeling batch effect in population-scale scRNA-seq data</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-celltype-prop-modeling.html">Model cell type proportions for new individuals</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL-NL.html">Model nonlinear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL.html">Model linear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-modify-eQTL-effects.html">Modify eQTL effect for eGenes / non-eGenes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-ES-specification.html">Power analysis based on user specified eQTL effect sizes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-fitted.html">Power analysis based on a fitted scDesignPop marginal model</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-selected.html">Power analysis for selected genes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-trans-eQTL.html">Modeling trans-eQTL effect using public data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/chrisycd/scDesignPop/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Model nonlinear dynamic eQTL effects in continuous cell states</h1>
                        <h4 data-toc-skip class="author">Chris Dong</h4>
            <address class="author_afil">
      Department of Statistics and Data Science, University of
California, Los
Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:cycd@g.ucla.edu" class="email">cycd@g.ucla.edu</a>
      </address>
                              <h4 data-toc-skip class="author">Yihui
Cen</h4>
            <address class="author_afil">
      Department of Computational Medicine, University of California,
Los Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:yihuicen@g.ucla.edu" class="email">yihuicen@g.ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">23 January 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chrisycd/scDesignPop/blob/HEAD/vignettes/scDesignPop-dynamic-eQTL-NL.Rmd"><code>vignettes/scDesignPop-dynamic-eQTL-NL.Rmd</code></a></small>
      <div class="d-none name"><code>scDesignPop-dynamic-eQTL-NL.Rmd</code></div>
    </div>

    
    
<style type="text/css">
pre {
  white-space: pre !important;
  overflow-x: scroll !important;
}
</style>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chrisycd/scDesignPop">scDesignPop</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="step-1-construct-a-data-list">Step 1: construct a data list<a class="anchor" aria-label="anchor" href="#step-1-construct-a-data-list"></a>
</h2>
<p>scDesignPop can also be extended to other eQTL effects. Here we show
how scDesignPop can also model nonlinear dynamic eQTL effects in
continuous cell states to mimick the real data better. To run
scDesignPop, a list of data is required as input. This is done using the
<code>constructDataPop</code> function. A
<code>SingleCellExperiment</code> object and an <code>eqtlgeno</code>
dataframe are the two main inputs needed. The <code>eqtlgeno</code>
dataframe consists of eQTL annotations (it must have cell state, gene,
SNP, chromosome, and position columns at a minimum), and genotypes
across individuals (columns) for every SNP (rows). The structure of an
example <code>eqtlgeno</code> dataframe is given below. Here, we use a
subset of the B cells from OneK1K cohort as example.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_sce_Bcell"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_eqtlgeno_Bcell"</span><span class="op">)</span></span>
<span><span class="va">col_Bcell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">example_sce_Bcell</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">example_sce_Bcell</span><span class="op">)</span><span class="op">$</span><span class="va">slingPseudotime_1_sq</span> <span class="op">&lt;-</span> <span class="va">col_Bcell</span><span class="op">$</span><span class="va">slingPseudotime_1</span><span class="op">^</span><span class="fl">2</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/constructDataPop.html">constructDataPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce_Bcell</span>,</span>
<span>    eqtlgeno_df <span class="op">=</span> <span class="va">example_eqtlgeno_Bcell</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">example_sce_Bcell</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    overlap_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    sampid_vec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_variable <span class="op">=</span> <span class="st">"slingPseudotime_1"</span>,</span>
<span>    n_quantiles <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    slot_name <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    snp_model <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>    time_colname <span class="op">=</span> <span class="st">"slingPseudotime_1"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    feature_colname <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    chrom_colname <span class="op">=</span> <span class="st">"CHR"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    prune_thres <span class="op">=</span> <span class="fl">0.9</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2-fit-marginal-model">Step 2: fit marginal model<a class="anchor" aria-label="anchor" href="#step-2-fit-marginal-model"></a>
</h2>
<p>Next, a marginal model is specified to fit each gene using the
<code>fitMarginalPop</code> function.<br>
Here we use a Negative Binominal as the parametric model using
<code>"nb"</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitMarginalPop.html">fitMarginalPop</a></span><span class="op">(</span></span>
<span>    data_list <span class="op">=</span> <span class="va">data_list</span>,</span>
<span>    mean_formula <span class="op">=</span> <span class="st">"(1|indiv) + slingPseudotime_1 + slingPseudotime_1_sq"</span>,</span>
<span>    model_family <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    interact_colnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slingPseudotime_1"</span>, <span class="st">"slingPseudotime_1_sq"</span><span class="op">)</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcapply"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    filter_snps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    snpvar_thres <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    force_formula <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    data_maxsize <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-3-fit-a-gaussian-copula">Step 3: fit a Gaussian copula<a class="anchor" aria-label="anchor" href="#step-3-fit-a-gaussian-copula"></a>
</h2>
<p>The third step is to fit a Gaussian copula using the
<code>fitCopulaPop</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span>, kind <span class="op">=</span> <span class="st">"L'Ecuyer-CMRG"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">copula_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitCopulaPop.html">fitCopulaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce_Bcell</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    copula <span class="op">=</span> <span class="st">"gaussian"</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"mcmapply"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGkind</a></span><span class="op">(</span><span class="st">"Mersenne-Twister"</span><span class="op">)</span>  <span class="co"># reset</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-4-extract-parameters">Step 4: extract parameters<a class="anchor" aria-label="anchor" href="#step-4-extract-parameters"></a>
</h2>
<p>The fourth step is to compute the mean, sigma, and zero probability
parameters using the <code>extractParaPop</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">para_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractParaPop.html">extractParaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce_Bcell</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_eqtl_geno_list <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"eqtl_geno_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-5-simulate-counts">Step 5: simulate counts<a class="anchor" aria-label="anchor" href="#step-5-simulate-counts"></a>
</h2>
<p>The fifth step is to simulate counts using the
<code>simuNewPop</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">newcount_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simuNewPop.html">simuNewPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce_Bcell</span>,</span>
<span>    mean_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"mean_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    sigma_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"sigma_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    zero_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"zero_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    quantile_mat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_list <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"copula_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    nonnegative <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    important_feature <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"important_feature"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    filtered_gene <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"filtered_gene"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-6-create-singlecellexperiment-object-using-simulated-data">Step 6: create SingleCellExperiment object using simulated data<a class="anchor" aria-label="anchor" href="#step-6-create-singlecellexperiment-object-using-simulated-data"></a>
</h2>
<p>After simulating the data, we can create a
<code>SingleCellExperiment</code> object as follows.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simu_sce_Bcell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">newcount_mat</span><span class="op">)</span>, </span>
<span>                                       colData <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">simu_sce_Bcell</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"counts"</span></span>
<span></span>
<span><span class="co"># save gene meta data</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">simu_sce_Bcell</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">example_sce_Bcell</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-7-visualize-using-phate">Step 7: visualize using PHATE<a class="anchor" aria-label="anchor" href="#step-7-visualize-using-phate"></a>
</h2>
<p>The simulated data can be visualized using a UMAP plot as
follows.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">phateR</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run PHATE for reference data</span></span>
<span><span class="va">ph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phateR/man/phate.html" class="external-link">phate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">example_sce_Bcell</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating PHATE...</span></span>
<span><span class="co">#&gt;   Running PHATE on 3726 observations and 817 variables.</span></span>
<span><span class="co">#&gt;   Calculating graph and diffusion operator...</span></span>
<span><span class="co">#&gt;     Calculating PCA...</span></span>
<span><span class="co">#&gt;     Calculated PCA in 0.50 seconds.</span></span>
<span><span class="co">#&gt;     Calculating KNN search...</span></span>
<span><span class="co">#&gt;     Calculated KNN search in 1.03 seconds.</span></span>
<span><span class="co">#&gt;     Calculating affinities...</span></span>
<span><span class="co">#&gt;     Calculated affinities in 0.03 seconds.</span></span>
<span><span class="co">#&gt;   Calculated graph and diffusion operator in 1.57 seconds.</span></span>
<span><span class="co">#&gt;   Calculating landmark operator...</span></span>
<span><span class="co">#&gt;     Calculating SVD...</span></span>
<span><span class="co">#&gt;     Calculated SVD in 0.45 seconds.</span></span>
<span><span class="co">#&gt;     Calculating KMeans...</span></span>
<span><span class="co">#&gt;     Calculated KMeans in 8.25 seconds.</span></span>
<span><span class="co">#&gt;   Calculated landmark operator in 9.23 seconds.</span></span>
<span><span class="co">#&gt;   Calculating optimal t...</span></span>
<span><span class="co">#&gt;     Automatically selected t = 18</span></span>
<span><span class="co">#&gt;   Calculated optimal t in 1.09 seconds.</span></span>
<span><span class="co">#&gt;   Calculating diffusion potential...</span></span>
<span><span class="co">#&gt;   Calculated diffusion potential in 0.22 seconds.</span></span>
<span><span class="co">#&gt;   Calculating metric MDS...</span></span>
<span><span class="co">#&gt;   Calculated metric MDS in 4.57 seconds.</span></span>
<span><span class="co">#&gt; Calculated PHATE in 16.68 seconds.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">example_sce_Bcell</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">SimpleList</span><span class="op">(</span>PHATE <span class="op">=</span> <span class="va">ph</span><span class="op">$</span><span class="va">embedding</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run PHATE for simulated data</span></span>
<span><span class="va">ph_simu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phateR/man/phate.html" class="external-link">phate</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simu_sce_Bcell</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating PHATE...</span></span>
<span><span class="co">#&gt;   Running PHATE on 3726 observations and 817 variables.</span></span>
<span><span class="co">#&gt;   Calculating graph and diffusion operator...</span></span>
<span><span class="co">#&gt;     Calculating PCA...</span></span>
<span><span class="co">#&gt;     Calculated PCA in 0.45 seconds.</span></span>
<span><span class="co">#&gt;     Calculating KNN search...</span></span>
<span><span class="co">#&gt;     Calculated KNN search in 1.01 seconds.</span></span>
<span><span class="co">#&gt;     Calculating affinities...</span></span>
<span><span class="co">#&gt;     Calculated affinities in 0.02 seconds.</span></span>
<span><span class="co">#&gt;   Calculated graph and diffusion operator in 1.48 seconds.</span></span>
<span><span class="co">#&gt;   Calculating landmark operator...</span></span>
<span><span class="co">#&gt;     Calculating SVD...</span></span>
<span><span class="co">#&gt;     Calculated SVD in 0.32 seconds.</span></span>
<span><span class="co">#&gt;     Calculating KMeans...</span></span>
<span><span class="co">#&gt;     Calculated KMeans in 8.19 seconds.</span></span>
<span><span class="co">#&gt;   Calculated landmark operator in 9.14 seconds.</span></span>
<span><span class="co">#&gt;   Calculating optimal t...</span></span>
<span><span class="co">#&gt;     Automatically selected t = 15</span></span>
<span><span class="co">#&gt;   Calculated optimal t in 1.13 seconds.</span></span>
<span><span class="co">#&gt;   Calculating diffusion potential...</span></span>
<span><span class="co">#&gt;   Calculated diffusion potential in 0.23 seconds.</span></span>
<span><span class="co">#&gt;   Calculating metric MDS...</span></span>
<span><span class="co">#&gt;   Calculated metric MDS in 4.37 seconds.</span></span>
<span><span class="co">#&gt; Calculated PHATE in 16.34 seconds.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">simu_sce_Bcell</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">SimpleList</span><span class="op">(</span>PHATE <span class="op">=</span> <span class="va">ph_simu</span><span class="op">$</span><span class="va">embedding</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">example_sce_Bcell</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">PHATE1</span>, <span class="va">PHATE2</span>, color <span class="op">=</span> <span class="va">example_sce_Bcell</span><span class="op">$</span><span class="va">slingPseudotime_1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggrastr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggrastr/man/rasterise.html" class="external-link">rasterise</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    dpi <span class="op">=</span> <span class="fl">300</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Pseudotime"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Reference"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_simu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">simu_sce_Bcell</span><span class="op">)</span>, </span>
<span>                 <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">PHATE1</span>, <span class="va">PHATE2</span>, </span>
<span>                     color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">simu_sce_Bcell</span><span class="op">)</span><span class="op">$</span><span class="va">slingPseudotime_1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggrastr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggrastr/man/rasterise.html" class="external-link">rasterise</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    dpi <span class="op">=</span> <span class="fl">300</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"Pseudotime"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"scDesignPop (Nonlinear Dynamic eQTL)"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">p_simu</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDesignPop-dynamic-eQTL-NL_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] phateR_1.0.7                Matrix_1.6-5               </span></span>
<span><span class="co">#&gt;  [3] scater_1.26.1               ggplot2_3.5.2              </span></span>
<span><span class="co">#&gt;  [5] scuttle_1.8.4               SingleCellExperiment_1.20.1</span></span>
<span><span class="co">#&gt;  [7] SummarizedExperiment_1.28.0 Biobase_2.58.0             </span></span>
<span><span class="co">#&gt;  [9] GenomicRanges_1.50.2        GenomeInfoDb_1.34.9        </span></span>
<span><span class="co">#&gt; [11] IRanges_2.32.0              S4Vectors_0.36.2           </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.44.0         MatrixGenerics_1.10.0      </span></span>
<span><span class="co">#&gt; [15] matrixStats_1.1.0           scDesignPop_0.0.0.9009     </span></span>
<span><span class="co">#&gt; [17] BiocStyle_2.26.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] ggbeeswarm_0.7.2          minqa_1.2.8              </span></span>
<span><span class="co">#&gt;   [3] rprojroot_2.1.1           XVector_0.38.0           </span></span>
<span><span class="co">#&gt;   [5] BiocNeighbors_1.16.0      fs_1.6.6                 </span></span>
<span><span class="co">#&gt;   [7] rstudioapi_0.17.1         glmmTMB_1.1.13           </span></span>
<span><span class="co">#&gt;   [9] farver_2.1.2              ggrepel_0.9.5            </span></span>
<span><span class="co">#&gt;  [11] mvtnorm_1.3-3             codetools_0.2-20         </span></span>
<span><span class="co">#&gt;  [13] splines_4.2.3             sparseMatrixStats_1.10.0 </span></span>
<span><span class="co">#&gt;  [15] cachem_1.1.0              knitr_1.50               </span></span>
<span><span class="co">#&gt;  [17] jsonlite_2.0.0            nloptr_2.2.1             </span></span>
<span><span class="co">#&gt;  [19] Cairo_1.6-2               png_0.1-8                </span></span>
<span><span class="co">#&gt;  [21] BiocManager_1.30.25       compiler_4.2.3           </span></span>
<span><span class="co">#&gt;  [23] assertthat_0.2.1          fastmap_1.2.0            </span></span>
<span><span class="co">#&gt;  [25] cli_3.6.5                 BiocSingular_1.14.0      </span></span>
<span><span class="co">#&gt;  [27] htmltools_0.5.8.1         tools_4.2.3              </span></span>
<span><span class="co">#&gt;  [29] rsvd_1.0.5                gtable_0.3.6             </span></span>
<span><span class="co">#&gt;  [31] glue_1.8.0                GenomeInfoDbData_1.2.9   </span></span>
<span><span class="co">#&gt;  [33] dplyr_1.1.4               rappdirs_0.3.3           </span></span>
<span><span class="co">#&gt;  [35] Rcpp_1.0.14               jquerylib_0.1.4          </span></span>
<span><span class="co">#&gt;  [37] pkgdown_2.2.0             vctrs_0.6.5              </span></span>
<span><span class="co">#&gt;  [39] nlme_3.1-164              DelayedMatrixStats_1.20.0</span></span>
<span><span class="co">#&gt;  [41] xfun_0.52                 rbibutils_2.3            </span></span>
<span><span class="co">#&gt;  [43] beachmat_2.14.2           lme4_1.1-35.3            </span></span>
<span><span class="co">#&gt;  [45] lifecycle_1.0.4           irlba_2.3.5.1            </span></span>
<span><span class="co">#&gt;  [47] zlibbioc_1.44.0           MASS_7.3-58.2            </span></span>
<span><span class="co">#&gt;  [49] zoo_1.8-14                scales_1.4.0             </span></span>
<span><span class="co">#&gt;  [51] ragg_1.5.0                parallel_4.2.3           </span></span>
<span><span class="co">#&gt;  [53] sandwich_3.1-1            TMB_1.9.11               </span></span>
<span><span class="co">#&gt;  [55] RColorBrewer_1.1-3        yaml_2.3.10              </span></span>
<span><span class="co">#&gt;  [57] memoise_2.0.1             reticulate_1.42.0        </span></span>
<span><span class="co">#&gt;  [59] pbapply_1.7-2             gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [61] ggrastr_1.0.2             sass_0.4.10              </span></span>
<span><span class="co">#&gt;  [63] desc_1.4.3                ScaledMatrix_1.6.0       </span></span>
<span><span class="co">#&gt;  [65] boot_1.3-30               BiocParallel_1.32.6      </span></span>
<span><span class="co">#&gt;  [67] Rdpack_2.6.4              rlang_1.1.6              </span></span>
<span><span class="co">#&gt;  [69] pkgconfig_2.0.3           systemfonts_1.2.3        </span></span>
<span><span class="co">#&gt;  [71] bitops_1.0-9              evaluate_1.0.3           </span></span>
<span><span class="co">#&gt;  [73] lattice_0.22-6            patchwork_1.2.0          </span></span>
<span><span class="co">#&gt;  [75] htmlwidgets_1.6.4         labeling_0.4.3           </span></span>
<span><span class="co">#&gt;  [77] tidyselect_1.2.1          here_1.0.1               </span></span>
<span><span class="co">#&gt;  [79] magrittr_2.0.3            bookdown_0.43            </span></span>
<span><span class="co">#&gt;  [81] R6_2.6.1                  reformulas_0.4.1         </span></span>
<span><span class="co">#&gt;  [83] generics_0.1.4            DelayedArray_0.24.0      </span></span>
<span><span class="co">#&gt;  [85] pillar_1.10.2             withr_3.0.2              </span></span>
<span><span class="co">#&gt;  [87] mgcv_1.9-1                RCurl_1.98-1.17          </span></span>
<span><span class="co">#&gt;  [89] tibble_3.2.1              rmarkdown_2.27           </span></span>
<span><span class="co">#&gt;  [91] viridis_0.6.5             grid_4.2.3               </span></span>
<span><span class="co">#&gt;  [93] digest_0.6.37             pbmcapply_1.5.1          </span></span>
<span><span class="co">#&gt;  [95] numDeriv_2016.8-1.1       textshaping_0.4.0        </span></span>
<span><span class="co">#&gt;  [97] beeswarm_0.4.0            viridisLite_0.4.2        </span></span>
<span><span class="co">#&gt;  [99] vipor_0.4.7               bslib_0.9.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Y. Dong, Yihui Cen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
