<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Power analysis based on user specified eQTL effect sizes â€¢ scDesignPop</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Power analysis based on user specified eQTL effect sizes">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDesignPop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9010</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDesignPop.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/scDesignPop-batch-effect.html">Modeling batch effect in population-scale scRNA-seq data</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-celltype-prop-modeling.html">Model cell type proportions for new individuals</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL-NL.html">Model nonlinear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL.html">Model linear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-modify-eQTL-effects.html">Modify eQTL effect for eGenes / non-eGenes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-ES-specification.html">Power analysis based on user specified eQTL effect sizes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-selected.html">Power analysis for selected genes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-trans-eQTL.html">Modeling trans-eQTL effect using public data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/chrisycd/scDesignPop/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Power analysis based on user specified eQTL effect sizes</h1>
                        <h4 data-toc-skip class="author">Chris Dong</h4>
            <address class="author_afil">
      Department of Statistics and Data Science, University of
California, Los
Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:cycd@g.ucla.edu" class="email">cycd@g.ucla.edu</a>
      </address>
                              <h4 data-toc-skip class="author">Yihui
Cen</h4>
            <address class="author_afil">
      Department of Computational Medicine, University of California,
Los Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:yihuicen@g.ucla.edu" class="email">yihuicen@g.ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">13 February 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chrisycd/scDesignPop/blob/HEAD/vignettes/scDesignPop-power-analysis-ES-specification.Rmd"><code>vignettes/scDesignPop-power-analysis-ES-specification.Rmd</code></a></small>
      <div class="d-none name"><code>scDesignPop-power-analysis-ES-specification.Rmd</code></div>
    </div>

    
    
<style type="text/css">
pre {
  white-space: pre !important;
  overflow-x: scroll !important;
}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>scDesignPop provides power analysis tools at cell-type-specific
level. Users can specify cell-type-specific eQTL effect sizes to check
hypothetical scenarios and how power curve will change based on eQTL
effect size specifications.</p>
</div>
<div class="section level2">
<h2 id="library-and-data-preparation">Library and data preparation<a class="anchor" aria-label="anchor" href="#library-and-data-preparation"></a>
</h2>
<p>For data preparation, a list of data is required as input. This is
done using the <code>constructDataPop</code> function. A
<code>SingleCellExperiment</code> object and an <code>eqtlgeno</code>
dataframe are the two main inputs needed. The <code>eqtlgeno</code>
dataframe consists of eQTL annotations (it must have cell state, gene,
SNP, chromosome, and position columns at a minimum), and genotypes
across individuals (columns) for every SNP (rows). The structure of an
example <code>eqtlgeno</code> dataframe is given below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chrisycd/scDesignPop">scDesignPop</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_sce"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_eqtlgeno"</span><span class="op">)</span></span>
<span><span class="va">example_sce_sel</span> <span class="op">&lt;-</span> <span class="va">example_sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000163221"</span>,<span class="st">"ENSG00000135218"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">example_eqtlgeno_sel</span> <span class="op">&lt;-</span> <span class="va">example_eqtlgeno</span><span class="op">[</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">example_eqtlgeno</span><span class="op">$</span><span class="va">gene_id</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000163221"</span>,<span class="st">"ENSG00000135218"</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_list_sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/constructDataPop.html">constructDataPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce_sel</span>,</span>
<span>    eqtlgeno_df <span class="op">=</span> <span class="va">example_eqtlgeno_sel</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">example_sce_sel</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    overlap_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    sampid_vec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_variable <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    slot_name <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    snp_mode <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    feature_colname <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    chrom_colname <span class="op">=</span> <span class="st">"CHR"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    prune_thres <span class="op">=</span> <span class="fl">0.9</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fitting-the-marginal-model">Fitting the marginal model<a class="anchor" aria-label="anchor" href="#fitting-the-marginal-model"></a>
</h2>
<p>Next, a marginal model is specified to fit each gene using the
<code>fitMarginalPop</code> function.<br>
Here we use a Negative Binominal as the parametric model using
<code>"nb"</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list_sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitMarginalPop.html">fitMarginalPop</a></span><span class="op">(</span></span>
<span>    data_list <span class="op">=</span> <span class="va">data_list_sel</span>,</span>
<span>    mean_formula <span class="op">=</span> <span class="st">"(1|indiv) + cell_type"</span>,</span>
<span>    model_family <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    interact_colnames <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcapply"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    filter_snps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    snpvar_thres <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    force_formula <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    data_maxsize <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="performing-power-analysis-with-user-specified-eqtl-effect-size-at-cell-type-level">Performing power analysis with user specified eQTL effect size at
cell-type level<a class="anchor" aria-label="anchor" href="#performing-power-analysis-with-user-specified-eqtl-effect-size-at-cell-type-level"></a>
</h2>
<p>Given fitted marginal model, scDesignPop can perform simulation-based
power analysis for a specific gene-SNP pair across selected cell types
using the <code>runPowerAnalysis</code> function. Based on the previous
naming of covariates, we specify the fitted snpid as
<code>"1:153337943"</code>, the name of the column for fixed cell state
effect and random individual effect as <code>"cell_type"</code> and
<code>"indiv"</code> in the input parameters. To check these namings, we
can call the covariate data frame using
<code>marginal_list_sel[["ENSG00000163221"]]$fit$frame</code>. The
selected cell types for testing are specified in cellstate_vector and
have to be consistent with the covariate data frame.</p>
<p>Particarly, parameters <code>snp_number</code> and
<code>gene_number</code> are used to account for multiple testing
correction with Bonferroni correction. Parameter <code>methods</code> is
used to specify the marginal eQTL model from
<code>c("nb", "poisson", "gaussian", "pseudoBulkLinear")</code>.
Parameter <code>nindivs</code> and <code>ncells</code> are used to
specify the number of individuals and number of cells per individual,
from which we can analyze the performance of power analysis and find the
optimal setting. Here, we set <code>power_nsim = 1000</code> to increase
the number of simulations so we can calculate power with a higher
resolution. Using <code>power_nsim = 100</code> in default or smaller
values can reduce the computation time cost.</p>
<p>Specifically, users can specify different eQTL effect sizes in each
cell type using the input parameter
<code>celltype_specific_ES_list</code> as follows, where each element of
the list represents one scenario allowing eQTL effect sizes in multiple
cell types can be specified at the same time. (bmem cell type is also
included but without user specification)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">power_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPowerAnalysis.html">runPowerAnalysis</a></span><span class="op">(</span>marginal_list <span class="op">=</span> <span class="va">marginal_list_sel</span>,</span>
<span>                               marginal_model <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>                               geneid <span class="op">=</span> <span class="st">"ENSG00000163221"</span>,</span>
<span>                               snpid <span class="op">=</span> <span class="st">"1:153337943"</span>,</span>
<span>                               celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>                               celltype_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"monoc"</span><span class="op">)</span>,</span>
<span>                               celltype_specific_ES_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"monoc"</span> <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"monoc"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"monoc"</span> <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                               indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>                               methods <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"poisson"</span>,<span class="st">"pseudoBulkLinear"</span><span class="op">)</span>,</span>
<span>                               nindivs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">200</span><span class="op">)</span>,</span>
<span>                               ncells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">50</span><span class="op">)</span>,</span>
<span>                               alpha <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                               power_nsim <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                               snp_number <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                               gene_number <span class="op">=</span> <span class="fl">800</span>,</span>
<span>                               CI_nsim <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                               CI_conf <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                               ncores <span class="op">=</span> <span class="fl">25</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1.803924</span></span>
<span><span class="co">#&gt; [1] 0.05</span></span>
<span><span class="co">#&gt; [1] 1.803924</span></span>
<span><span class="co">#&gt; [1] 0.1</span></span>
<span><span class="co">#&gt; [1] 1.803924</span></span>
<span><span class="co">#&gt; [1] 0.2</span></span>
<span><span class="co">#&gt; [1] 1.803924</span></span>
<span><span class="co">#&gt; [1] 0.05</span></span>
<span><span class="co">#&gt; [1] 1.803924</span></span>
<span><span class="co">#&gt; [1] 0.1</span></span>
<span><span class="co">#&gt; [1] 1.803924</span></span>
<span><span class="co">#&gt; [1] 0.2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization-of-power-results">Visualization of power results<a class="anchor" aria-label="anchor" href="#visualization-of-power-results"></a>
</h2>
<p>The power analysis results can be visualized using the
<code>visualizePowerCurve</code> function. The cell type names in the
<code>cellstate_vector</code> must be included in the above power
analysis.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualizePowerCurve.html">visualizePowerCurve</a></span><span class="op">(</span>power_result <span class="op">=</span> <span class="va">power_data</span>,</span>
<span>                    celltype_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"monoc"</span><span class="op">)</span>,</span>
<span>                    x_axis <span class="op">=</span> <span class="st">"specifiedES"</span>,</span>
<span>                    x_facet<span class="op">=</span><span class="st">"nindiv"</span>,</span>
<span>                    y_facet <span class="op">=</span> <span class="st">"ncell"</span>,</span>
<span>                    col_group <span class="op">=</span> <span class="st">"method"</span>,</span>
<span>                    geneid <span class="op">=</span> <span class="st">"ENSG00000163221"</span>,</span>
<span>                    snpid <span class="op">=</span> <span class="st">"1:153337943"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDesignPop-power-analysis-ES-specification_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" width="700"></p>
<p>By swaping the x and y axis, we can show the result in a different
way.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualizePowerCurve.html">visualizePowerCurve</a></span><span class="op">(</span>power_result <span class="op">=</span> <span class="va">power_data</span>,</span>
<span>                    celltype_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"monoc"</span><span class="op">)</span>,</span>
<span>                    x_axis <span class="op">=</span> <span class="st">"specifiedES"</span>,</span>
<span>                    x_facet<span class="op">=</span><span class="st">"ncell"</span>,</span>
<span>                    y_facet <span class="op">=</span> <span class="st">"nindiv"</span>,</span>
<span>                    col_group <span class="op">=</span> <span class="st">"method"</span>,</span>
<span>                    geneid <span class="op">=</span> <span class="st">"ENSG00000163221"</span>,</span>
<span>                    snpid <span class="op">=</span> <span class="st">"1:153337943"</span><span class="op">)</span></span></code></pre></div>
<p><img src="scDesignPop-power-analysis-ES-specification_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Y. Dong, Yihui Cen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
