<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Modeling batch effect in population-scale scRNA-seq data â€¢ scDesignPop</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Modeling batch effect in population-scale scRNA-seq data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDesignPop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9009</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDesignPop.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/scDesignPop-batch-effect.html">Modeling batch effect in population-scale scRNA-seq data</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-celltype-prop-modeling.html">Model cell type proportions for new individuals</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL-NL.html">Model nonlinear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL.html">Model linear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-modify-eQTL-effects.html">Modify eQTL effect for eGenes / non-eGenes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-ES-specification.html">Power analysis based on user specified eQTL effect sizes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-fitted.html">Power analysis based on a fitted scDesignPop marginal model</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-selected.html">Power analysis for selected genes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-trans-eQTL.html">Modeling trans-eQTL effect using public data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/chrisycd/scDesignPop/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Modeling batch effect in population-scale scRNA-seq data</h1>
                        <h4 data-toc-skip class="author">Chris Dong</h4>
            <address class="author_afil">
      Department of Statistics and Data Science, University of
California, Los
Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:cycd@g.ucla.edu" class="email">cycd@g.ucla.edu</a>
      </address>
                              <h4 data-toc-skip class="author">Yihui
Cen</h4>
            <address class="author_afil">
      Department of Computational Medicine, University of California,
Los Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:yihuicen@g.ucla.edu" class="email">yihuicen@g.ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">23 January 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chrisycd/scDesignPop/blob/HEAD/vignettes/scDesignPop-batch-effect.Rmd"><code>vignettes/scDesignPop-batch-effect.Rmd</code></a></small>
      <div class="d-none name"><code>scDesignPop-batch-effect.Rmd</code></div>
    </div>

    
    
<style type="text/css">
pre {
  white-space: pre !important;
  overflow-x: scroll !important;
}
</style>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chrisycd/scDesignPop">scDesignPop</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>scDesignPop can also model population-scale scRNA-seq data with batch
information. The detailed tutorial is shown below.</p>
</div>
<div class="section level2">
<h2 id="fitting-batch-effects">Fitting batch effects<a class="anchor" aria-label="anchor" href="#fitting-batch-effects"></a>
</h2>
<div class="section level3">
<h3 id="step-1-construct-a-data-list">Step 1: construct a data list<a class="anchor" aria-label="anchor" href="#step-1-construct-a-data-list"></a>
</h3>
<p>To run scDesignPop, a list of data is required as input. This is done
using the <code>constructDataPop</code> function. A
<code>SingleCellExperiment</code> object and an <code>eqtlgeno</code>
dataframe are the two main inputs needed. The <code>eqtlgeno</code>
dataframe consists of eQTL annotations (it must have cell state, gene,
SNP, chromosome, and position columns at a minimum), and genotypes
across individuals (columns) for every SNP (rows). The structure of an
example <code>eqtlgeno</code> dataframe is given below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_sce"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_eqtlgeno"</span><span class="op">)</span></span></code></pre></div>
<p>Here, the <code>example_sce</code> also has a batch column.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 7811 rows and 5 columns</span></span>
<span><span class="co">#&gt;            indiv cell_type       sex       age    batch</span></span>
<span><span class="co">#&gt;      &lt;character&gt;  &lt;factor&gt; &lt;integer&gt; &lt;integer&gt; &lt;factor&gt;</span></span>
<span><span class="co">#&gt; 1          SAMP1     cd4nc         2        65        3</span></span>
<span><span class="co">#&gt; 2          SAMP1     cd4nc         2        65        3</span></span>
<span><span class="co">#&gt; 3          SAMP1     cd4nc         2        65        3</span></span>
<span><span class="co">#&gt; 4          SAMP1     cd4nc         2        65        3</span></span>
<span><span class="co">#&gt; 5          SAMP1     cd4nc         2        65        3</span></span>
<span><span class="co">#&gt; ...          ...       ...       ...       ...      ...</span></span>
<span><span class="co">#&gt; 7807      SAMP14     monoc         1        87        1</span></span>
<span><span class="co">#&gt; 7808      SAMP14     bmem          1        87        1</span></span>
<span><span class="co">#&gt; 7809      SAMP14     bmem          1        87        1</span></span>
<span><span class="co">#&gt; 7810      SAMP14     bmem          1        87        1</span></span>
<span><span class="co">#&gt; 7811      SAMP14     bmem          1        87        1</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/constructDataPop.html">constructDataPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    eqtlgeno_df <span class="op">=</span> <span class="va">example_eqtlgeno</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    overlap_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    sampid_vec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_variable <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    slot_name <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    snp_model <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    feature_colname <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    chrom_colname <span class="op">=</span> <span class="st">"CHR"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    prune_thres <span class="op">=</span> <span class="fl">0.9</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-fit-marginal-model">Step 2: fit marginal model<a class="anchor" aria-label="anchor" href="#step-2-fit-marginal-model"></a>
</h3>
<p>Next, a marginal model is specified to fit each gene using the
<code>fitMarginalPop</code> function.<br>
Here we use a Negative Binominal as the parametric model using
<code>"nb"</code>. Since we want to include the batch effect modeling,
we specify
<code>mean_formula = "(1|indiv) + cell_type + batch</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitMarginalPop.html">fitMarginalPop</a></span><span class="op">(</span></span>
<span>    data_list <span class="op">=</span> <span class="va">data_list</span>,</span>
<span>    mean_formula <span class="op">=</span> <span class="st">"(1|indiv) + cell_type + batch"</span>,</span>
<span>    model_family <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    interact_colnames <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcapply"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    filter_snps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    snpvar_thres <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    force_formula <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    data_maxsize <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-3-fit-a-gaussian-copula">Step 3: fit a Gaussian copula<a class="anchor" aria-label="anchor" href="#step-3-fit-a-gaussian-copula"></a>
</h3>
<p>The third step is to fit a Gaussian copula using the
<code>fitCopulaPop</code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span>, kind <span class="op">=</span> <span class="st">"L'Ecuyer-CMRG"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">copula_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitCopulaPop.html">fitCopulaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    copula <span class="op">=</span> <span class="st">"gaussian"</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"mcmapply"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGkind</a></span><span class="op">(</span><span class="st">"Mersenne-Twister"</span><span class="op">)</span>  <span class="co"># reset</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-4-extract-parameters">Step 4: extract parameters<a class="anchor" aria-label="anchor" href="#step-4-extract-parameters"></a>
</h3>
<p>The fourth step is to compute the mean, sigma, and zero probability
parameters using the <code>extractParaPop</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">para_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractParaPop.html">extractParaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_eqtl_geno_list <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"eqtl_geno_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-5-simulate-counts">Step 5: simulate counts<a class="anchor" aria-label="anchor" href="#step-5-simulate-counts"></a>
</h3>
<p>The fifth step is to simulate counts using the
<code>simuNewPop</code> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">newcount_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simuNewPop.html">simuNewPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    mean_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"mean_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    sigma_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"sigma_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    zero_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"zero_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    quantile_mat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_list <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"copula_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    nonnegative <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    important_feature <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"important_feature"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    filtered_gene <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"filtered_gene"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="removing-the-batch-effects">Removing the batch effects<a class="anchor" aria-label="anchor" href="#removing-the-batch-effects"></a>
</h2>
<div class="section level3">
<h3 id="step-1-set-zero-batch-effect-sizes-in-marginal-models">Step 1: set zero batch effect sizes in marginal models<a class="anchor" aria-label="anchor" href="#step-1-set-zero-batch-effect-sizes-in-marginal-models"></a>
</h3>
<p>Check the rank of the batch effect sizes in the marginal
distributions. Here, we take the gene <code>ENSG00000023902</code> as
example since every gene are fitted using the same formula.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list</span><span class="op">$</span><span class="va">ENSG00000023902</span><span class="op">$</span><span class="va">fit</span></span>
<span><span class="co">#&gt; Formula:          response ~ (1 | indiv) + cell_type + batch + `1:150133323` +  </span></span>
<span><span class="co">#&gt;     `1:150133323`:cell_type</span></span>
<span><span class="co">#&gt; Data: res_list[["dmat_df"]]</span></span>
<span><span class="co">#&gt;       AIC       BIC    logLik -2*log(L)  df.resid </span></span>
<span><span class="co">#&gt;  8048.368  8138.891 -4011.184  8022.368      7798 </span></span>
<span><span class="co">#&gt; Random-effects (co)variances:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;  Groups Name        Std.Dev.</span></span>
<span><span class="co">#&gt;  indiv  (Intercept) 0.1807  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of obs: 7811 / Conditional model: indiv, 40</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Dispersion parameter for nbinom2 family (): 1.63 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fixed Effects:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;                   (Intercept)                cell_typemononc  </span></span>
<span><span class="co">#&gt;                      -0.87229                        0.33417  </span></span>
<span><span class="co">#&gt;                 cell_typebmem                 cell_typecd4nc  </span></span>
<span><span class="co">#&gt;                      -0.69781                       -1.65119  </span></span>
<span><span class="co">#&gt;                        batch2                         batch3  </span></span>
<span><span class="co">#&gt;                      -0.34148                       -0.12831  </span></span>
<span><span class="co">#&gt;                        batch4                  `1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.11636                        0.18264  </span></span>
<span><span class="co">#&gt; cell_typemononc:`1:150133323`    cell_typebmem:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.16217                       -0.04084  </span></span>
<span><span class="co">#&gt;  cell_typecd4nc:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.11806</span></span></code></pre></div>
<p>After the ranks of the batch effects are obtained, set zero to the
gene marginal models and check the summary report again.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">marginal_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">par</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>  <span class="va">x</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">parfull</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>  <span class="va">x</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">sdr</span><span class="op">$</span><span class="va">par.fixed</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>  <span class="va">x</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">marginal_list_null</span><span class="op">$</span><span class="va">ENSG00000023902</span><span class="op">$</span><span class="va">fit</span></span>
<span><span class="co">#&gt; Formula:          response ~ (1 | indiv) + cell_type + batch + `1:150133323` +  </span></span>
<span><span class="co">#&gt;     `1:150133323`:cell_type</span></span>
<span><span class="co">#&gt; Data: res_list[["dmat_df"]]</span></span>
<span><span class="co">#&gt;       AIC       BIC    logLik -2*log(L)  df.resid </span></span>
<span><span class="co">#&gt;  8048.368  8138.891 -4011.184  8022.368      7798 </span></span>
<span><span class="co">#&gt; Random-effects (co)variances:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;  Groups Name        Std.Dev.</span></span>
<span><span class="co">#&gt;  indiv  (Intercept) 0.1807  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of obs: 7811 / Conditional model: indiv, 40</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Dispersion parameter for nbinom2 family (): 1.63 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fixed Effects:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;                   (Intercept)                cell_typemononc  </span></span>
<span><span class="co">#&gt;                      -0.87229                        0.33417  </span></span>
<span><span class="co">#&gt;                 cell_typebmem                 cell_typecd4nc  </span></span>
<span><span class="co">#&gt;                      -0.69781                       -1.65119  </span></span>
<span><span class="co">#&gt;                        batch2                         batch3  </span></span>
<span><span class="co">#&gt;                       0.00000                        0.00000  </span></span>
<span><span class="co">#&gt;                        batch4                  `1:150133323`  </span></span>
<span><span class="co">#&gt;                       0.00000                        0.18264  </span></span>
<span><span class="co">#&gt; cell_typemononc:`1:150133323`    cell_typebmem:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.16217                       -0.04084  </span></span>
<span><span class="co">#&gt;  cell_typecd4nc:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.11806</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-re-extract-the-parameters">Step 2: re-extract the parameters<a class="anchor" aria-label="anchor" href="#step-2-re-extract-the-parameters"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">para_new_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractParaPop.html">extractParaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list_null</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_eqtl_geno_list <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"eqtl_geno_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-3-simulate-counts-with-no-batch-effects">Step 3: simulate counts with no batch effects<a class="anchor" aria-label="anchor" href="#step-3-simulate-counts-with-no-batch-effects"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">newcount_mat_null</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simuNewPop.html">simuNewPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    mean_mat <span class="op">=</span> <span class="va">para_new_null</span><span class="op">[[</span><span class="st">"mean_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    sigma_mat <span class="op">=</span> <span class="va">para_new_null</span><span class="op">[[</span><span class="st">"sigma_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    zero_mat <span class="op">=</span> <span class="va">para_new_null</span><span class="op">[[</span><span class="st">"zero_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    quantile_mat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_list <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"copula_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    nonnegative <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    important_feature <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"important_feature"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    filtered_gene <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"filtered_gene"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="set-new-batch-effects">Set new batch effects<a class="anchor" aria-label="anchor" href="#set-new-batch-effects"></a>
</h2>
<div class="section level3">
<h3 id="step-1-set-new-batch-effect-sizes-in-marginal-models">Step 1: set new batch effect sizes in marginal models<a class="anchor" aria-label="anchor" href="#step-1-set-new-batch-effect-sizes-in-marginal-models"></a>
</h3>
<p>Check the rank of the batch effect sizes in the marginal
distributions. Here, we take the gene <code>ENSG00000023902</code> as
example since every gene are fitted using the same formula.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list</span><span class="op">$</span><span class="va">ENSG00000023902</span><span class="op">$</span><span class="va">fit</span></span>
<span><span class="co">#&gt; Formula:          response ~ (1 | indiv) + cell_type + batch + `1:150133323` +  </span></span>
<span><span class="co">#&gt;     `1:150133323`:cell_type</span></span>
<span><span class="co">#&gt; Data: res_list[["dmat_df"]]</span></span>
<span><span class="co">#&gt;       AIC       BIC    logLik -2*log(L)  df.resid </span></span>
<span><span class="co">#&gt;  8048.368  8138.891 -4011.184  8022.368      7798 </span></span>
<span><span class="co">#&gt; Random-effects (co)variances:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;  Groups Name        Std.Dev.</span></span>
<span><span class="co">#&gt;  indiv  (Intercept) 0.1807  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of obs: 7811 / Conditional model: indiv, 40</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Dispersion parameter for nbinom2 family (): 1.63 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fixed Effects:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;                   (Intercept)                cell_typemononc  </span></span>
<span><span class="co">#&gt;                      -0.87229                        0.33417  </span></span>
<span><span class="co">#&gt;                 cell_typebmem                 cell_typecd4nc  </span></span>
<span><span class="co">#&gt;                      -0.69781                       -1.65119  </span></span>
<span><span class="co">#&gt;                        batch2                         batch3  </span></span>
<span><span class="co">#&gt;                      -0.34148                       -0.12831  </span></span>
<span><span class="co">#&gt;                        batch4                  `1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.11636                        0.18264  </span></span>
<span><span class="co">#&gt; cell_typemononc:`1:150133323`    cell_typebmem:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.16217                       -0.04084  </span></span>
<span><span class="co">#&gt;  cell_typecd4nc:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.11806</span></span></code></pre></div>
<p>After the ranks of the batch effects are obtained, set a new value to
the gene marginal models and check the summary report again.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list_diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">marginal_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">val</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">3</span>, mean <span class="op">=</span> <span class="fl">5</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="va">x</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">par</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">val</span></span>
<span>  <span class="va">x</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">parfull</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">val</span></span>
<span>  <span class="va">x</span><span class="op">$</span><span class="va">fit</span><span class="op">$</span><span class="va">sdr</span><span class="op">$</span><span class="va">par.fixed</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">val</span></span>
<span>  <span class="va">x</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">marginal_list_diff</span><span class="op">$</span><span class="va">ENSG00000023902</span><span class="op">$</span><span class="va">fit</span></span>
<span><span class="co">#&gt; Formula:          response ~ (1 | indiv) + cell_type + batch + `1:150133323` +  </span></span>
<span><span class="co">#&gt;     `1:150133323`:cell_type</span></span>
<span><span class="co">#&gt; Data: res_list[["dmat_df"]]</span></span>
<span><span class="co">#&gt;       AIC       BIC    logLik -2*log(L)  df.resid </span></span>
<span><span class="co">#&gt;  8048.368  8138.891 -4011.184  8022.368      7798 </span></span>
<span><span class="co">#&gt; Random-effects (co)variances:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;  Groups Name        Std.Dev.</span></span>
<span><span class="co">#&gt;  indiv  (Intercept) 0.1807  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of obs: 7811 / Conditional model: indiv, 40</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Dispersion parameter for nbinom2 family (): 1.63 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fixed Effects:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Conditional model:</span></span>
<span><span class="co">#&gt;                   (Intercept)                cell_typemononc  </span></span>
<span><span class="co">#&gt;                      -0.87229                        0.33417  </span></span>
<span><span class="co">#&gt;                 cell_typebmem                 cell_typecd4nc  </span></span>
<span><span class="co">#&gt;                      -0.69781                       -1.65119  </span></span>
<span><span class="co">#&gt;                        batch2                         batch3  </span></span>
<span><span class="co">#&gt;                       1.00619                        5.68744  </span></span>
<span><span class="co">#&gt;                        batch4                  `1:150133323`  </span></span>
<span><span class="co">#&gt;                       3.15227                        0.18264  </span></span>
<span><span class="co">#&gt; cell_typemononc:`1:150133323`    cell_typebmem:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.16217                       -0.04084  </span></span>
<span><span class="co">#&gt;  cell_typecd4nc:`1:150133323`  </span></span>
<span><span class="co">#&gt;                      -0.11806</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-re-extract-the-parameters-1">Step 2: re-extract the parameters<a class="anchor" aria-label="anchor" href="#step-2-re-extract-the-parameters-1"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">para_new_diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractParaPop.html">extractParaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list_diff</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_eqtl_geno_list <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"eqtl_geno_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-3-simulate-counts-with-no-batch-effects-1">Step 3: simulate counts with no batch effects<a class="anchor" aria-label="anchor" href="#step-3-simulate-counts-with-no-batch-effects-1"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">newcount_mat_diff</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simuNewPop.html">simuNewPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    mean_mat <span class="op">=</span> <span class="va">para_new_diff</span><span class="op">[[</span><span class="st">"mean_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    sigma_mat <span class="op">=</span> <span class="va">para_new_diff</span><span class="op">[[</span><span class="st">"sigma_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    zero_mat <span class="op">=</span> <span class="va">para_new_diff</span><span class="op">[[</span><span class="st">"zero_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    quantile_mat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_list <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"copula_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    nonnegative <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    important_feature <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"important_feature"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    filtered_gene <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"filtered_gene"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] ggplot2_3.5.2               SingleCellExperiment_1.20.1</span></span>
<span><span class="co">#&gt;  [3] SummarizedExperiment_1.28.0 Biobase_2.58.0             </span></span>
<span><span class="co">#&gt;  [5] GenomicRanges_1.50.2        GenomeInfoDb_1.34.9        </span></span>
<span><span class="co">#&gt;  [7] IRanges_2.32.0              S4Vectors_0.36.2           </span></span>
<span><span class="co">#&gt;  [9] BiocGenerics_0.44.0         MatrixGenerics_1.10.0      </span></span>
<span><span class="co">#&gt; [11] matrixStats_1.1.0           scDesignPop_0.0.0.9009     </span></span>
<span><span class="co">#&gt; [13] BiocStyle_2.26.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] sass_0.4.10            jsonlite_2.0.0         splines_4.2.3         </span></span>
<span><span class="co">#&gt;  [4] bslib_0.9.0            Rdpack_2.6.4           assertthat_0.2.1      </span></span>
<span><span class="co">#&gt;  [7] BiocManager_1.30.25    GenomeInfoDbData_1.2.9 yaml_2.3.10           </span></span>
<span><span class="co">#&gt; [10] numDeriv_2016.8-1.1    pillar_1.10.2          lattice_0.22-6        </span></span>
<span><span class="co">#&gt; [13] glue_1.8.0             reformulas_0.4.1       digest_0.6.37         </span></span>
<span><span class="co">#&gt; [16] RColorBrewer_1.1-3     XVector_0.38.0         rbibutils_2.3         </span></span>
<span><span class="co">#&gt; [19] glmmTMB_1.1.13         minqa_1.2.8            sandwich_3.1-1        </span></span>
<span><span class="co">#&gt; [22] htmltools_0.5.8.1      Matrix_1.6-5           pkgconfig_2.0.3       </span></span>
<span><span class="co">#&gt; [25] bookdown_0.43          zlibbioc_1.44.0        mvtnorm_1.3-3         </span></span>
<span><span class="co">#&gt; [28] scales_1.4.0           lme4_1.1-35.3          tibble_3.2.1          </span></span>
<span><span class="co">#&gt; [31] mgcv_1.9-1             generics_0.1.4         farver_2.1.2          </span></span>
<span><span class="co">#&gt; [34] cachem_1.1.0           withr_3.0.2            pbapply_1.7-2         </span></span>
<span><span class="co">#&gt; [37] TMB_1.9.11             cli_3.6.5              magrittr_2.0.3        </span></span>
<span><span class="co">#&gt; [40] evaluate_1.0.3         fs_1.6.6               nlme_3.1-164          </span></span>
<span><span class="co">#&gt; [43] MASS_7.3-58.2          textshaping_0.4.0      tools_4.2.3           </span></span>
<span><span class="co">#&gt; [46] lifecycle_1.0.4        DelayedArray_0.24.0    compiler_4.2.3        </span></span>
<span><span class="co">#&gt; [49] pkgdown_2.2.0          jquerylib_0.1.4        pbmcapply_1.5.1       </span></span>
<span><span class="co">#&gt; [52] systemfonts_1.2.3      rlang_1.1.6            grid_4.2.3            </span></span>
<span><span class="co">#&gt; [55] RCurl_1.98-1.17        nloptr_2.2.1           rstudioapi_0.17.1     </span></span>
<span><span class="co">#&gt; [58] htmlwidgets_1.6.4      bitops_1.0-9           rmarkdown_2.27        </span></span>
<span><span class="co">#&gt; [61] boot_1.3-30            gtable_0.3.6           R6_2.6.1              </span></span>
<span><span class="co">#&gt; [64] zoo_1.8-14             knitr_1.50             dplyr_1.1.4           </span></span>
<span><span class="co">#&gt; [67] fastmap_1.2.0          ragg_1.5.0             desc_1.4.3            </span></span>
<span><span class="co">#&gt; [70] parallel_4.2.3         Rcpp_1.0.14            vctrs_0.6.5           </span></span>
<span><span class="co">#&gt; [73] tidyselect_1.2.1       xfun_0.52</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Y. Dong, Yihui Cen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
