<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Modeling trans-eQTL effect using public data • scDesignPop</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Modeling trans-eQTL effect using public data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDesignPop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9010</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/scDesignPop.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/scDesignPop-batch-effect.html">Modeling batch effect in population-scale scRNA-seq data</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-celltype-prop-modeling.html">Model cell type proportions for new individuals</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL-NL.html">Model nonlinear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL.html">Model linear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-modify-eQTL-effects.html">Modify eQTL effect for eGenes / non-eGenes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-ES-specification.html">Power analysis based on user specified eQTL effect sizes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-selected.html">Power analysis for selected genes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-trans-eQTL.html">Modeling trans-eQTL effect using public data</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/chrisycd/scDesignPop/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Modeling trans-eQTL effect using public data</h1>
                        <h4 data-toc-skip class="author">Chris Dong</h4>
            <address class="author_afil">
      Department of Statistics and Data Science, University of
California, Los
Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:cycd@g.ucla.edu" class="email">cycd@g.ucla.edu</a>
      </address>
                              <h4 data-toc-skip class="author">Yihui
Cen</h4>
            <address class="author_afil">
      Department of Computational Medicine, University of California,
Los Angeles<br><a class="author_email" href="mailto:#"></a><a href="mailto:yihuicen@g.ucla.edu" class="email">yihuicen@g.ucla.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">13 February 2026</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chrisycd/scDesignPop/blob/HEAD/vignettes/scDesignPop-trans-eQTL.Rmd"><code>vignettes/scDesignPop-trans-eQTL.Rmd</code></a></small>
      <div class="d-none name"><code>scDesignPop-trans-eQTL.Rmd</code></div>
    </div>

    
    
<style type="text/css">
pre {
  white-space: pre !important;
  overflow-x: scroll !important;
}
</style>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chrisycd/scDesignPop">scDesignPop</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>scDesignPop supports user-defined eQTL pairs, making it
straightforward to extend the framework to trans-eQTL studies with known
ground truth.</p>
</div>
<div class="section level2">
<h2 id="library-and-data-preparation">Library and data preparation<a class="anchor" aria-label="anchor" href="#library-and-data-preparation"></a>
</h2>
<div class="section level3">
<h3 id="prepare-the-eqtl-genotype-dataframe-as-input">prepare the eQTL genotype dataframe as input<a class="anchor" aria-label="anchor" href="#prepare-the-eqtl-genotype-dataframe-as-input"></a>
</h3>
<p>Here, we take the trans-eQTL list from eQTLGen phase 1 (<a href="https://www.eqtlgen.org/trans-eqtls.html" class="external-link uri">https://www.eqtlgen.org/trans-eqtls.html</a>) as an example
to show the scDesignPop’s flexibility in eQTL ground truth settings. The
download link of the trans-eQTL list is <a href="https://download.gcc.rug.nl/downloads/eqtlgen/trans-eqtl/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz" class="external-link uri">https://download.gcc.rug.nl/downloads/eqtlgen/trans-eqtl/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz</a>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">wget</span> https://download.gcc.rug.nl/downloads/eqtlgen/trans-eqtl/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co">#&gt; --2026-02-13 01:24:58--  https://download.gcc.rug.nl/downloads/eqtlgen/trans-eqtl/2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co">#&gt; Resolving download.gcc.rug.nl (download.gcc.rug.nl)... 195.169.22.66</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co">#&gt; Connecting to download.gcc.rug.nl (download.gcc.rug.nl)|195.169.22.66|:443... connected.</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co">#&gt; HTTP request sent, awaiting response... 200 OK</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="co">#&gt; Length: 1839918 (1.8M) [application/octet-stream]</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co">#&gt; Saving to: ‘2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz.12’</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a><span class="co">#&gt;      0K .......... .......... .......... .......... ..........  2%  177K 10s</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a><span class="co">#&gt;     50K .......... .......... .......... .......... ..........  5%  329K 7s</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co">#&gt;    100K .......... .......... .......... .......... ..........  8% 61.5M 5s</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="co">#&gt;    150K .......... .......... .......... .......... .......... 11% 2.37M 4s</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a><span class="co">#&gt;    200K .......... .......... .......... .......... .......... 13%  382K 4s</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a><span class="co">#&gt;    250K .......... .......... .......... .......... .......... 16% 68.6M 3s</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a><span class="co">#&gt;    300K .......... .......... .......... .......... .......... 19% 85.3M 2s</span></span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a><span class="co">#&gt;    350K .......... .......... .......... .......... .......... 22%  107M 2s</span></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co">#&gt;    400K .......... .......... .......... .......... .......... 25% 2.48M 2s</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="co">#&gt;    450K .......... .......... .......... .......... .......... 27%  382K 2s</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a><span class="co">#&gt;    500K .......... .......... .......... .......... .......... 30% 84.8M 2s</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a><span class="co">#&gt;    550K .......... .......... .......... .......... .......... 33%  141M 1s</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a><span class="co">#&gt;    600K .......... .......... .......... .......... .......... 36% 84.5M 1s</span></span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a><span class="co">#&gt;    650K .......... .......... .......... .......... .......... 38%  103M 1s</span></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co">#&gt;    700K .......... .......... .......... .......... .......... 41%  112M 1s</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a><span class="co">#&gt;    750K .......... .......... .......... .......... .......... 44%  129M 1s</span></span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="co">#&gt;    800K .......... .......... .......... .......... .......... 47% 2.69M 1s</span></span>
<span id="cb2-26"><a href="#cb2-26" tabindex="-1"></a><span class="co">#&gt;    850K .......... .......... .......... .......... .......... 50%  122M 1s</span></span>
<span id="cb2-27"><a href="#cb2-27" tabindex="-1"></a><span class="co">#&gt;    900K .......... .......... .......... .......... .......... 52%  383K 1s</span></span>
<span id="cb2-28"><a href="#cb2-28" tabindex="-1"></a><span class="co">#&gt;    950K .......... .......... .......... .......... .......... 55%  123M 1s</span></span>
<span id="cb2-29"><a href="#cb2-29" tabindex="-1"></a><span class="co">#&gt;   1000K .......... .......... .......... .......... .......... 58% 89.0M 1s</span></span>
<span id="cb2-30"><a href="#cb2-30" tabindex="-1"></a><span class="co">#&gt;   1050K .......... .......... .......... .......... .......... 61%  136M 1s</span></span>
<span id="cb2-31"><a href="#cb2-31" tabindex="-1"></a><span class="co">#&gt;   1100K .......... .......... .......... .......... .......... 64%  131M 1s</span></span>
<span id="cb2-32"><a href="#cb2-32" tabindex="-1"></a><span class="co">#&gt;   1150K .......... .......... .......... .......... .......... 66% 94.2M 0s</span></span>
<span id="cb2-33"><a href="#cb2-33" tabindex="-1"></a><span class="co">#&gt;   1200K .......... .......... .......... .......... .......... 69%  119M 0s</span></span>
<span id="cb2-34"><a href="#cb2-34" tabindex="-1"></a><span class="co">#&gt;   1250K .......... .......... .......... .......... .......... 72%  136M 0s</span></span>
<span id="cb2-35"><a href="#cb2-35" tabindex="-1"></a><span class="co">#&gt;   1300K .......... .......... .......... .......... .......... 75%  130M 0s</span></span>
<span id="cb2-36"><a href="#cb2-36" tabindex="-1"></a><span class="co">#&gt;   1350K .......... .......... .......... .......... .......... 77% 80.2M 0s</span></span>
<span id="cb2-37"><a href="#cb2-37" tabindex="-1"></a><span class="co">#&gt;   1400K .......... .......... .......... .......... .......... 80%  143K 0s</span></span>
<span id="cb2-38"><a href="#cb2-38" tabindex="-1"></a><span class="co">#&gt;   1450K .......... .......... .......... .......... .......... 83%  110K 0s</span></span>
<span id="cb2-39"><a href="#cb2-39" tabindex="-1"></a><span class="co">#&gt;   1500K .......... .......... .......... .......... .......... 86%  329K 0s</span></span>
<span id="cb2-40"><a href="#cb2-40" tabindex="-1"></a><span class="co">#&gt;   1550K .......... .......... .......... .......... .......... 89%  329K 0s</span></span>
<span id="cb2-41"><a href="#cb2-41" tabindex="-1"></a><span class="co">#&gt;   1600K .......... .......... .......... .......... .......... 91%  329K 0s</span></span>
<span id="cb2-42"><a href="#cb2-42" tabindex="-1"></a><span class="co">#&gt;   1650K .......... .......... .......... .......... .......... 94%  329K 0s</span></span>
<span id="cb2-43"><a href="#cb2-43" tabindex="-1"></a><span class="co">#&gt;   1700K .......... .......... .......... .......... .......... 97% 93.6M 0s</span></span>
<span id="cb2-44"><a href="#cb2-44" tabindex="-1"></a><span class="co">#&gt;   1750K .......... .......... .......... .......... ......    100%  170M=2.3s</span></span>
<span id="cb2-45"><a href="#cb2-45" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb2-46"><a href="#cb2-46" tabindex="-1"></a><span class="co">#&gt; 2026-02-13 01:25:01 (778 KB/s) - ‘2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz.12’ saved [1839918/1839918]</span></span></code></pre></div>
<p>Then, we reformat this trans-eQTL list into the required eQTL
dataframe as follows,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chrisycd/scDesignPop">scDesignPop</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">transeqtl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/data.table/man/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="st">"2018-09-04-trans-eQTLsFDR0.05-CohortInfoRemoved-BonferroniAdded.txt.gz"</span><span class="op">)</span></span>
<span><span class="va">transeqtl</span> <span class="op">&lt;-</span> <span class="va">transeqtl</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene"</span>,<span class="st">"SNP"</span>,<span class="st">"SNPChr"</span>,<span class="st">"SNPPos"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">transeqtl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="st">"bin"</span>, <span class="va">transeqtl</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">transeqtl</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell_type"</span>,<span class="st">"gene_id"</span>,<span class="st">"snp_id"</span>,<span class="st">"CHR"</span>,<span class="st">"POS"</span><span class="op">)</span></span>
<span><span class="va">transeqtl</span> <span class="op">&lt;-</span> <span class="va">transeqtl</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">transeqtl</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">transeqtl</span><span class="op">$</span><span class="va">snp_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">transeqtl</span><span class="op">$</span><span class="va">CHR</span>,<span class="va">transeqtl</span><span class="op">$</span><span class="va">POS</span>,sep <span class="op">=</span> <span class="st">":"</span><span class="op">)</span></span></code></pre></div>
<p>Since the genotype data are not provided and are often under
restricted access, here we extract the SNPs that we have genotype data
in the <code>example_eqtlgeno</code></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_eqtlgeno"</span><span class="op">)</span></span>
<span><span class="va">transeqtl</span> <span class="op">&lt;-</span> <span class="va">transeqtl</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">transeqtl</span><span class="op">$</span><span class="va">snp_id</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="va">example_eqtlgeno</span><span class="op">$</span><span class="va">snp_id</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">geno_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^SAMP"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">example_eqtlgeno</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">geno_by_snp</span> <span class="op">&lt;-</span> <span class="va">example_eqtlgeno</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">example_eqtlgeno</span><span class="op">$</span><span class="va">snp_id</span><span class="op">)</span>, </span>
<span>                                <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"snp_id"</span>, <span class="va">geno_cols</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">transeqtl</span><span class="op">$</span><span class="va">snp_id</span>, <span class="va">geno_by_snp</span><span class="op">$</span><span class="va">snp_id</span><span class="op">)</span></span>
<span><span class="va">trans_eqtlgeno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="va">transeqtl</span>,</span>
<span>  <span class="va">geno_by_snp</span><span class="op">[</span><span class="va">idx</span>, <span class="va">geno_cols</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">trans_eqtlgeno</span></span>
<span><span class="co">#&gt;      cell_type         gene_id      snp_id   CHR       POS SAMP1 SAMP2 SAMP3</span></span>
<span><span class="co">#&gt;         &lt;char&gt;          &lt;char&gt;      &lt;char&gt; &lt;int&gt;     &lt;int&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">#&gt;   1:       bin ENSG00000172262 12:69732105    12  69732105     2     0     1</span></span>
<span><span class="co">#&gt;   2:       bin ENSG00000258730 12:69732105    12  69732105     2     0     1</span></span>
<span><span class="co">#&gt;   3:       bin ENSG00000117862   11:327143    11    327143     0     1     0</span></span>
<span><span class="co">#&gt;   4:       bin ENSG00000171101 10:94450233    10  94450233     1     2     1</span></span>
<span><span class="co">#&gt;   5:       bin ENSG00000197057 10:94450233    10  94450233     1     2     1</span></span>
<span><span class="co">#&gt;  ---                                                                        </span></span>
<span><span class="co">#&gt; 129:       bin ENSG00000138433 3:127474030     3 127474030     0     2     2</span></span>
<span><span class="co">#&gt; 130:       bin ENSG00000182405  5:64355060     5  64355060     2     2     2</span></span>
<span><span class="co">#&gt; 131:       bin ENSG00000105085 12:56401085    12  56401085     1     0     2</span></span>
<span><span class="co">#&gt; 132:       bin ENSG00000214955 3:194403578     3 194403578     0     1     1</span></span>
<span><span class="co">#&gt; 133:       bin ENSG00000120093 10:94450233    10  94450233     1     2     1</span></span>
<span><span class="co">#&gt;      SAMP4 SAMP5 SAMP6 SAMP7 SAMP8 SAMP9 SAMP10 SAMP11 SAMP12 SAMP13 SAMP14</span></span>
<span><span class="co">#&gt;      &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt; &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;</span></span>
<span><span class="co">#&gt;   1:     2     1     1     1     1     0      2      2      1      2      1</span></span>
<span><span class="co">#&gt;   2:     2     1     1     1     1     0      2      2      1      2      1</span></span>
<span><span class="co">#&gt;   3:     1     0     1     0     2     1      0      1      1      1      0</span></span>
<span><span class="co">#&gt;   4:     0     2     2     1     0     0      1      1      0      2      2</span></span>
<span><span class="co">#&gt;   5:     0     2     2     1     0     0      1      1      0      2      2</span></span>
<span><span class="co">#&gt;  ---                                                                       </span></span>
<span><span class="co">#&gt; 129:     2     2     0     1     2     0      2      2      2      2      2</span></span>
<span><span class="co">#&gt; 130:     0     2     2     1     1     2      1      0      0      0      1</span></span>
<span><span class="co">#&gt; 131:     2     1     0     2     1     1      1      1      0      1      1</span></span>
<span><span class="co">#&gt; 132:     2     2     0     1     0     1      0      1      1      0      1</span></span>
<span><span class="co">#&gt; 133:     0     2     2     1     0     0      1      1      0      2      2</span></span>
<span><span class="co">#&gt;      SAMP15 SAMP16 SAMP17 SAMP18 SAMP19 SAMP20 SAMP21 SAMP22 SAMP23 SAMP24</span></span>
<span><span class="co">#&gt;       &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;</span></span>
<span><span class="co">#&gt;   1:      1      1      0      1      0      1      2      1      1      1</span></span>
<span><span class="co">#&gt;   2:      1      1      0      1      0      1      2      1      1      1</span></span>
<span><span class="co">#&gt;   3:      0      0      1      0      0      1      0      0      0      0</span></span>
<span><span class="co">#&gt;   4:      1      0      1      0      0      2      1      2      1      0</span></span>
<span><span class="co">#&gt;   5:      1      0      1      0      0      2      1      2      1      0</span></span>
<span><span class="co">#&gt;  ---                                                                      </span></span>
<span><span class="co">#&gt; 129:      2      1      1      1      0      2      1      0      2      1</span></span>
<span><span class="co">#&gt; 130:      0      2      2      2      1      1      1      1      2      1</span></span>
<span><span class="co">#&gt; 131:      1      1      2      0      1      1      2      2      0      0</span></span>
<span><span class="co">#&gt; 132:      2      1      1      2      1      0      2      1      2      1</span></span>
<span><span class="co">#&gt; 133:      1      0      1      0      0      2      1      2      1      0</span></span>
<span><span class="co">#&gt;      SAMP25 SAMP26 SAMP27 SAMP28 SAMP29 SAMP30 SAMP31 SAMP32 SAMP33 SAMP34</span></span>
<span><span class="co">#&gt;       &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;</span></span>
<span><span class="co">#&gt;   1:      0      0      1      1      1      2      1      1      0      0</span></span>
<span><span class="co">#&gt;   2:      0      0      1      1      1      2      1      1      0      0</span></span>
<span><span class="co">#&gt;   3:      0      1      1      1      0      0      0      2      1      0</span></span>
<span><span class="co">#&gt;   4:      0      1      1      0      0      1      2      0      1      1</span></span>
<span><span class="co">#&gt;   5:      0      1      1      0      0      1      2      0      1      1</span></span>
<span><span class="co">#&gt;  ---                                                                      </span></span>
<span><span class="co">#&gt; 129:      2      1      2      0      1      1      2      1      2      1</span></span>
<span><span class="co">#&gt; 130:      2      0      2      1      0      0      2      2      0      1</span></span>
<span><span class="co">#&gt; 131:      1      2      2      0      2      0      2      2      1      2</span></span>
<span><span class="co">#&gt; 132:      1      2      2      1      1      0      0      0      0      0</span></span>
<span><span class="co">#&gt; 133:      0      1      1      0      0      1      2      0      1      1</span></span>
<span><span class="co">#&gt;      SAMP35 SAMP36 SAMP37 SAMP38 SAMP39 SAMP40</span></span>
<span><span class="co">#&gt;       &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;  &lt;num&gt;</span></span>
<span><span class="co">#&gt;   1:      1      1      1      1      2      1</span></span>
<span><span class="co">#&gt;   2:      1      1      1      1      2      1</span></span>
<span><span class="co">#&gt;   3:      0      0      0      0      0      0</span></span>
<span><span class="co">#&gt;   4:      1      1      0      0      2      2</span></span>
<span><span class="co">#&gt;   5:      1      1      0      0      2      2</span></span>
<span><span class="co">#&gt;  ---                                          </span></span>
<span><span class="co">#&gt; 129:      2      2      2      2      1      2</span></span>
<span><span class="co">#&gt; 130:      1      2      2      2      2      1</span></span>
<span><span class="co">#&gt; 131:      2      1      2      1      2      1</span></span>
<span><span class="co">#&gt; 132:      1      1      2      1      0      1</span></span>
<span><span class="co">#&gt; 133:      1      1      0      0      2      2</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="modeling-and-simulation">Modeling and simulation<a class="anchor" aria-label="anchor" href="#modeling-and-simulation"></a>
</h2>
<div class="section level3">
<h3 id="step-1-construct-a-data-list">Step 1: construct a data list<a class="anchor" aria-label="anchor" href="#step-1-construct-a-data-list"></a>
</h3>
<p>To run scDesignPop, a list of data is required as input. This is done
using the <code>constructDataPop</code> function. A
<code>SingleCellExperiment</code> object and an <code>eqtlgeno</code>
dataframe are the two main inputs needed. The <code>eqtlgeno</code>
dataframe consists of eQTL annotations (it must have cell state, gene,
SNP, chromosome, and position columns at a minimum), and genotypes
across individuals (columns) for every SNP (rows). Here, we use
<code>example_sce</code> to filter for overlapped genes in
`trans_eqtlgeno’ object created</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"example_sce"</span><span class="op">)</span></span>
<span><span class="va">overlap_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span>,<span class="va">trans_eqtlgeno</span><span class="op">$</span><span class="va">gene_id</span><span class="op">)</span></span>
<span><span class="va">trans_eqtlgeno</span> <span class="op">&lt;-</span> <span class="va">trans_eqtlgeno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">trans_eqtlgeno</span><span class="op">$</span><span class="va">gene_id</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="va">overlap_genes</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">example_sce</span> <span class="op">&lt;-</span> <span class="va">example_sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span><span class="va">overlap_genes</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/constructDataPop.html">constructDataPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    eqtlgeno_df <span class="op">=</span> <span class="va">trans_eqtlgeno</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    overlap_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    sampid_vec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_variable <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    slot_name <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    snp_mode <span class="op">=</span> <span class="st">"single"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    feature_colname <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    chrom_colname <span class="op">=</span> <span class="st">"CHR"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    prune_thres <span class="op">=</span> <span class="fl">0.9</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-2-fit-marginal-model">Step 2: fit marginal model<a class="anchor" aria-label="anchor" href="#step-2-fit-marginal-model"></a>
</h3>
<p>Next, a marginal model is specified to fit each gene using the
<code>fitMarginalPop</code> function.<br>
Here we use a Negative Binominal as the parametric model using
<code>"nb"</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marginal_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitMarginalPop.html">fitMarginalPop</a></span><span class="op">(</span></span>
<span>    data_list <span class="op">=</span> <span class="va">data_list</span>,</span>
<span>    mean_formula <span class="op">=</span> <span class="st">"(1|indiv) + cell_type"</span>,</span>
<span>    model_family <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    interact_colnames <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcapply"</span>,</span>
<span>    n_threads <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>    loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>    snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>    celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    filter_snps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    snpvar_thres <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    force_formula <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    data_maxsize <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-3-fit-a-gaussian-copula">Step 3: fit a Gaussian copula<a class="anchor" aria-label="anchor" href="#step-3-fit-a-gaussian-copula"></a>
</h3>
<p>The third step is to fit a Gaussian copula using the
<code>fitCopulaPop</code> function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span>, kind <span class="op">=</span> <span class="st">"L'Ecuyer-CMRG"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">copula_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitCopulaPop.html">fitCopulaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    copula <span class="op">=</span> <span class="st">"gaussian"</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"mcmapply"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">RNGkind</a></span><span class="op">(</span><span class="st">"Mersenne-Twister"</span><span class="op">)</span>  <span class="co"># reset</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-4-extract-parameters">Step 4: extract parameters<a class="anchor" aria-label="anchor" href="#step-4-extract-parameters"></a>
</h3>
<p>The fourth step is to compute the mean, sigma, and zero probability
parameters using the <code>extractParaPop</code> function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">para_new</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractParaPop.html">extractParaPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    assay_use <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>    marginal_list <span class="op">=</span> <span class="va">marginal_list</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_eqtl_geno_list <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"eqtl_geno_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-5-simulate-counts">Step 5: simulate counts<a class="anchor" aria-label="anchor" href="#step-5-simulate-counts"></a>
</h3>
<p>The fifth step is to simulate counts using the
<code>simuNewPop</code> function.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="va">newcount_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simuNewPop.html">simuNewPop</a></span><span class="op">(</span></span>
<span>    sce <span class="op">=</span> <span class="va">example_sce</span>,</span>
<span>    mean_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"mean_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    sigma_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"sigma_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    zero_mat <span class="op">=</span> <span class="va">para_new</span><span class="op">[[</span><span class="st">"zero_mat"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    quantile_mat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    copula_list <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"copula_list"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>    family_use <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>    nonnegative <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    input_data <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    new_covariate <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    important_feature <span class="op">=</span> <span class="va">copula_fit</span><span class="op">[[</span><span class="st">"important_feature"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    filtered_gene <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"filtered_gene"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    parallelization <span class="op">=</span> <span class="st">"pbmcmapply"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-6-create-singlecellexperiment-object-using-simulated-data">Step 6: create SingleCellExperiment object using simulated data<a class="anchor" aria-label="anchor" href="#step-6-create-singlecellexperiment-object-using-simulated-data"></a>
</h3>
<p>After simulating the data, we can create a
<code>SingleCellExperiment</code> object as follows.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simu_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">newcount_mat</span><span class="op">)</span>, </span>
<span>                                 colData <span class="op">=</span> <span class="va">data_list</span><span class="op">[[</span><span class="st">"new_covariate"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">simu_sce</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"counts"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">simu_sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span>
<span><span class="va">simu_sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 22 7811 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(22): ENSG00000198574 ENSG00000189292 ... ENSG00000100079</span></span>
<span><span class="co">#&gt;   ENSG00000225783</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(6): indiv cell_type ... batch corr_group</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] data.table_1.17.4           ggplot2_3.5.2              </span></span>
<span><span class="co">#&gt;  [3] SingleCellExperiment_1.20.1 SummarizedExperiment_1.28.0</span></span>
<span><span class="co">#&gt;  [5] Biobase_2.58.0              GenomicRanges_1.50.2       </span></span>
<span><span class="co">#&gt;  [7] GenomeInfoDb_1.34.9         IRanges_2.32.0             </span></span>
<span><span class="co">#&gt;  [9] S4Vectors_0.36.2            BiocGenerics_0.44.0        </span></span>
<span><span class="co">#&gt; [11] MatrixGenerics_1.10.0       matrixStats_1.1.0          </span></span>
<span><span class="co">#&gt; [13] scDesignPop_0.0.0.9010      BiocStyle_2.26.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] viridis_0.6.5          sass_0.4.10            jsonlite_2.0.0        </span></span>
<span><span class="co">#&gt;  [4] viridisLite_0.4.2      splines_4.2.3          R.utils_2.13.0        </span></span>
<span><span class="co">#&gt;  [7] RhpcBLASctl_0.23-42    bslib_0.9.0            assertthat_0.2.1      </span></span>
<span><span class="co">#&gt; [10] BiocManager_1.30.25    GenomeInfoDbData_1.2.9 yaml_2.3.10           </span></span>
<span><span class="co">#&gt; [13] numDeriv_2016.8-1.1    pillar_1.10.2          lattice_0.22-6        </span></span>
<span><span class="co">#&gt; [16] glue_1.8.0             digest_0.6.37          RColorBrewer_1.1-3    </span></span>
<span><span class="co">#&gt; [19] XVector_0.38.0         glmmTMB_1.1.9          minqa_1.2.8           </span></span>
<span><span class="co">#&gt; [22] R.oo_1.27.1            htmltools_0.5.8.1      Matrix_1.6-5          </span></span>
<span><span class="co">#&gt; [25] pkgconfig_2.0.3        bookdown_0.43          zlibbioc_1.44.0       </span></span>
<span><span class="co">#&gt; [28] mvtnorm_1.3-3          scales_1.4.0           lme4_1.1-35.3         </span></span>
<span><span class="co">#&gt; [31] tibble_3.2.1           mgcv_1.9-1             generics_0.1.4        </span></span>
<span><span class="co">#&gt; [34] farver_2.1.2           withr_3.0.2            cachem_1.1.0          </span></span>
<span><span class="co">#&gt; [37] pbapply_1.7-2          TMB_1.9.11             cli_3.6.5             </span></span>
<span><span class="co">#&gt; [40] magrittr_2.0.3         evaluate_1.0.3         R.methodsS3_1.8.2     </span></span>
<span><span class="co">#&gt; [43] fs_1.6.6               nlme_3.1-164           MASS_7.3-58.2         </span></span>
<span><span class="co">#&gt; [46] textshaping_0.4.0      tools_4.2.3            lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt; [49] DelayedArray_0.24.0    irlba_2.3.5.1          compiler_4.2.3        </span></span>
<span><span class="co">#&gt; [52] pkgdown_2.2.0          jquerylib_0.1.4        pbmcapply_1.5.1       </span></span>
<span><span class="co">#&gt; [55] systemfonts_1.2.3      rlang_1.1.6            grid_4.2.3            </span></span>
<span><span class="co">#&gt; [58] RCurl_1.98-1.17        nloptr_2.2.1           rstudioapi_0.17.1     </span></span>
<span><span class="co">#&gt; [61] htmlwidgets_1.6.4      bitops_1.0-9           rmarkdown_2.27        </span></span>
<span><span class="co">#&gt; [64] boot_1.3-30            gtable_0.3.6           R6_2.6.1              </span></span>
<span><span class="co">#&gt; [67] gridExtra_2.3          knitr_1.50             dplyr_1.1.4           </span></span>
<span><span class="co">#&gt; [70] fastmap_1.2.0          uwot_0.2.3             ragg_1.5.0            </span></span>
<span><span class="co">#&gt; [73] desc_1.4.3             parallel_4.2.3         Rcpp_1.0.14           </span></span>
<span><span class="co">#&gt; [76] vctrs_0.6.5            tidyselect_1.2.1       xfun_0.52</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Y. Dong, Yihui Cen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
