<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Construct a list of input data — constructDataPop • scDesignPop</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Construct a list of input data — constructDataPop"><meta name="description" content="Function extracts an expression matrix, cell covariates, and filters the SNPs
in the eQTL genotype dataframe."><meta property="og:description" content="Function extracts an expression matrix, cell covariates, and filters the SNPs
in the eQTL genotype dataframe."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDesignPop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9010</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scDesignPop.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/scDesignPop-batch-effect.html">Modeling batch effect in population-scale scRNA-seq data</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-celltype-prop-modeling.html">Model cell type proportions for new individuals</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL-NL.html">Model nonlinear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL.html">Model linear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-modify-eQTL-effects.html">Modify eQTL effect for eGenes / non-eGenes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-ES-specification.html">Power analysis based on user specified eQTL effect sizes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-selected.html">Power analysis for selected genes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-trans-eQTL.html">Modeling trans-eQTL effect using public data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/chrisycd/scDesignPop/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Construct a list of input data</h1>
      <small class="dont-index">Source: <a href="https://github.com/chrisycd/scDesignPop/blob/HEAD/R/constructDataPop.R"><code>R/constructDataPop.R</code></a></small>
      <div class="d-none name"><code>constructDataPop.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Function extracts an expression matrix, cell covariates, and filters the SNPs
in the eQTL genotype dataframe.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">constructDataPop</span><span class="op">(</span></span>
<span>  <span class="va">sce</span>,</span>
<span>  <span class="va">eqtlgeno_df</span>,</span>
<span>  new_eqtlgeno_df <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new_covariate <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  overlap_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampid_vec <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  copula_variable <span class="op">=</span> <span class="va">celltype_colname</span>,</span>
<span>  n_quantiles <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  slot_name <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  snp_mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"single"</span>, <span class="st">"multi"</span><span class="op">)</span>,</span>
<span>  time_colname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  feature_colname <span class="op">=</span> <span class="st">"gene_id"</span>,</span>
<span>  snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>  loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>  chrom_colname <span class="op">=</span> <span class="st">"CHR"</span>,</span>
<span>  indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>  prune_thres <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sce">sce<a class="anchor" aria-label="anchor" href="#arg-sce"></a></dt>
<dd><p>a SingleCellExperiment object.</p></dd>


<dt id="arg-eqtlgeno-df">eqtlgeno_df<a class="anchor" aria-label="anchor" href="#arg-eqtlgeno-df"></a></dt>
<dd><p>a dataframe with eQTL annotations and SNP genotypes for each gene.</p></dd>


<dt id="arg-new-eqtlgeno-df">new_eqtlgeno_df<a class="anchor" aria-label="anchor" href="#arg-new-eqtlgeno-df"></a></dt>
<dd><p>a dataframe with eQTL annotations and SNP genotypes for each
gene in new individuals. The default is NULL.</p></dd>


<dt id="arg-new-covariate">new_covariate<a class="anchor" aria-label="anchor" href="#arg-new-covariate"></a></dt>
<dd><p>a cell covariate dataframe for which to simulate.
The default is NULL.</p></dd>


<dt id="arg-overlap-features">overlap_features<a class="anchor" aria-label="anchor" href="#arg-overlap-features"></a></dt>
<dd><p>an optional string vector to filter for features (ie. genes).
The default is NULL.</p></dd>


<dt id="arg-sampid-vec">sampid_vec<a class="anchor" aria-label="anchor" href="#arg-sampid-vec"></a></dt>
<dd><p>an optional string vector to filter for sample ids.
The default is NULL.</p></dd>


<dt id="arg-copula-variable">copula_variable<a class="anchor" aria-label="anchor" href="#arg-copula-variable"></a></dt>
<dd><p>a string scalar specifying the column name in cell
covariate of <code>sce</code> object for which to fit the copula by. If NULL, then
copula will be fitted for all cells. if it's the same as celltype_colname,
then copula will be fitted per cell type, while if it's the same as
time_colname, copula will be fitted per time quantile.</p></dd>


<dt id="arg-n-quantiles">n_quantiles<a class="anchor" aria-label="anchor" href="#arg-n-quantiles"></a></dt>
<dd><p>an integer value used for how many time quantiles of continuous
cell states will be assigned for copula fitting. The default is 10.</p></dd>


<dt id="arg-slot-name">slot_name<a class="anchor" aria-label="anchor" href="#arg-slot-name"></a></dt>
<dd><p>a string scalar specifying the slot to use in input <code>sce</code>.
The default is "counts".</p></dd>


<dt id="arg-snp-mode">snp_mode<a class="anchor" aria-label="anchor" href="#arg-snp-mode"></a></dt>
<dd><p>a string scalar specifying the type of SNP mode used in the marginal
models. Options are either "single" for single-SNP mode, or "multi" for
multi-SNP mode. If multi-SNP mode is used, each gene will have one or more
SNPs in the model upon SNP pruning. The default is "single".</p></dd>


<dt id="arg-time-colname">time_colname<a class="anchor" aria-label="anchor" href="#arg-time-colname"></a></dt>
<dd><p>a string scalar specifying the time/pseudotime variable in
the cell covariate of <code>sce</code> object.</p></dd>


<dt id="arg-celltype-colname">celltype_colname<a class="anchor" aria-label="anchor" href="#arg-celltype-colname"></a></dt>
<dd><p>a string scalar specifying the cell type variable in
<code>eqtlgeno_df</code> and cell covariate of <code>sce</code> object.
The default is "cell_type".</p></dd>


<dt id="arg-feature-colname">feature_colname<a class="anchor" aria-label="anchor" href="#arg-feature-colname"></a></dt>
<dd><p>a string scalar specifying the feature variable (ie. genes)
in <code>eqtlgeno_df</code>. The default is "gene_id".</p></dd>


<dt id="arg-snp-colname">snp_colname<a class="anchor" aria-label="anchor" href="#arg-snp-colname"></a></dt>
<dd><p>a string scalar for the SNP variable in <code>eqtlgeno_df</code>.
The default is "snp_id".</p></dd>


<dt id="arg-loc-colname">loc_colname<a class="anchor" aria-label="anchor" href="#arg-loc-colname"></a></dt>
<dd><p>a string scalar for the last column of eQTL annotation in
<code>eqtlgeno_df</code>. The default is "POS".</p></dd>


<dt id="arg-chrom-colname">chrom_colname<a class="anchor" aria-label="anchor" href="#arg-chrom-colname"></a></dt>
<dd><p>a string scalar of the chromosome variable in <code>eqtlgeno_df</code>.
The default is "CHR".</p></dd>


<dt id="arg-indiv-colname">indiv_colname<a class="anchor" aria-label="anchor" href="#arg-indiv-colname"></a></dt>
<dd><p>a string scalar of the sample ID variable in cell covariate
of <code>sce</code>. The default is "indiv".</p></dd>


<dt id="arg-prune-thres">prune_thres<a class="anchor" aria-label="anchor" href="#arg-prune-thres"></a></dt>
<dd><p>a numerical value between 0 and 1 used to threshold the pairwise
correlations of eQTLs' genotypes for each feature. The default value is 0.9.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>additional arguments passed to internal functions.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>outputs a list with following elements:</p><dl><dt><code>count_mat</code></dt>
<dd><p>a cell-by-gene matrix of response values.</p></dd>

<dt><code>covariate</code></dt>
<dd><p>a cell-by-covariate data frame used for fit marginal.</p></dd>

<dt><code>new_covariate</code></dt>
<dd><p>an optional cell-by-covariate data frame used
in simulation.</p></dd>

<dt><code>important_features</code></dt>
<dd><p>a string vector of gene ids.</p></dd>

<dt><code>noeqtl_features</code></dt>
<dd><p>a string vector of gene ids that do not have eQTLs.</p></dd>

<dt><code>eqtl_geno_list</code></dt>
<dd><p>a list of eQTL genotype dataframes for each gene
(for fit marginal).</p></dd>

<dt><code>new_eqtl_geno_list</code></dt>
<dd><p>a optional list of eQTL genotype dataframes
for each gene (for new individual simulation).</p></dd>

<dt><code>filtered_gene</code></dt>
<dd><p>string vector of features QC filtered.</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="cn">NULL</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Y. Dong, Yihui Cen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

