<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Fit marginal models for every feature — fitMarginalPop • scDesignPop</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Fit marginal models for every feature — fitMarginalPop"><meta name="description" content="Fits a specified parametric model using various input parameters."><meta property="og:description" content="Fits a specified parametric model using various input parameters."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scDesignPop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9010</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/scDesignPop.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/scDesignPop-batch-effect.html">Modeling batch effect in population-scale scRNA-seq data</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-celltype-prop-modeling.html">Model cell type proportions for new individuals</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL-NL.html">Model nonlinear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-dynamic-eQTL.html">Model linear dynamic eQTL effects in continuous cell states</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-modify-eQTL-effects.html">Modify eQTL effect for eGenes / non-eGenes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-ES-specification.html">Power analysis based on user specified eQTL effect sizes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-power-analysis-selected.html">Power analysis for selected genes</a></li>
    <li><a class="dropdown-item" href="../articles/scDesignPop-trans-eQTL.html">Modeling trans-eQTL effect using public data</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/chrisycd/scDesignPop/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Fit marginal models for every feature</h1>
      <small class="dont-index">Source: <a href="https://github.com/chrisycd/scDesignPop/blob/HEAD/R/fitMarginalPop.R"><code>R/fitMarginalPop.R</code></a></small>
      <div class="d-none name"><code>fitMarginalPop.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Fits a specified parametric model using various input parameters.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fitMarginalPop</span><span class="op">(</span></span>
<span>  <span class="va">data_list</span>,</span>
<span>  <span class="va">mean_formula</span>,</span>
<span>  model_family <span class="op">=</span> <span class="st">"nb"</span>,</span>
<span>  interact_colnames <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  parallelization <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pbmcapply"</span>, <span class="st">"future.apply"</span>, <span class="st">"parallel"</span>, <span class="st">"biocparallel"</span><span class="op">)</span>,</span>
<span>  n_cores <span class="op">=</span> <span class="fl">2L</span>,</span>
<span>  loc_colname <span class="op">=</span> <span class="st">"POS"</span>,</span>
<span>  snp_colname <span class="op">=</span> <span class="st">"snp_id"</span>,</span>
<span>  celltype_colname <span class="op">=</span> <span class="st">"cell_type"</span>,</span>
<span>  indiv_colname <span class="op">=</span> <span class="st">"indiv"</span>,</span>
<span>  filter_snps <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  snpvar_thres <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  force_formula <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  keep_cellnames <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  future.seed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  data_maxsize <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-list">data_list<a class="anchor" aria-label="anchor" href="#arg-data-list"></a></dt>
<dd><p>a list of input data.</p></dd>


<dt id="arg-mean-formula">mean_formula<a class="anchor" aria-label="anchor" href="#arg-mean-formula"></a></dt>
<dd><p>a string scalar to specify the mean formula, including
random effects (if any) but without SNP genotypes or SNP genotype interaction effects.</p></dd>


<dt id="arg-model-family">model_family<a class="anchor" aria-label="anchor" href="#arg-model-family"></a></dt>
<dd><p>a string scalar to specify model fitting used.</p></dd>


<dt id="arg-interact-colnames">interact_colnames<a class="anchor" aria-label="anchor" href="#arg-interact-colnames"></a></dt>
<dd><p>a string scalar or vector for the variable names that
have first-order interaction with SNP genotypes.</p></dd>


<dt id="arg-parallelization">parallelization<a class="anchor" aria-label="anchor" href="#arg-parallelization"></a></dt>
<dd><p>a string scalar specifying the parallelization backend
used during marginal fitting. Must be one of either "parallel", "future.apply",
"biocparallel", or "pbmcapply". The default value is "pbmcapply". See details.</p></dd>


<dt id="arg-n-cores">n_cores<a class="anchor" aria-label="anchor" href="#arg-n-cores"></a></dt>
<dd><p>positive integer value (greater or equal to 1) to specify the
number of CPU cores used in parallelization. The default is 2.</p></dd>


<dt id="arg-loc-colname">loc_colname<a class="anchor" aria-label="anchor" href="#arg-loc-colname"></a></dt>
<dd><p>a string scalar for column name of SNP position variable.</p></dd>


<dt id="arg-snp-colname">snp_colname<a class="anchor" aria-label="anchor" href="#arg-snp-colname"></a></dt>
<dd><p>a string scalar for column name of SNP id variable.</p></dd>


<dt id="arg-celltype-colname">celltype_colname<a class="anchor" aria-label="anchor" href="#arg-celltype-colname"></a></dt>
<dd><p>a string scalar for column name of cell type.</p></dd>


<dt id="arg-indiv-colname">indiv_colname<a class="anchor" aria-label="anchor" href="#arg-indiv-colname"></a></dt>
<dd><p>a string scalar for column name of individuals (samples).</p></dd>


<dt id="arg-filter-snps">filter_snps<a class="anchor" aria-label="anchor" href="#arg-filter-snps"></a></dt>
<dd><p>a logical scalar for whether to filter out SNP covariates
with either low-variance or with only 1 distinct genotype (ie. all 1's)
prior to fitting the model.</p></dd>


<dt id="arg-snpvar-thres">snpvar_thres<a class="anchor" aria-label="anchor" href="#arg-snpvar-thres"></a></dt>
<dd><p>a numeric scalar (between 0 to 1) used to filter out SNPs
whose variance of genotypes across samples are below this threshold.
Used together when <code>filter_snps = TRUE</code>.</p></dd>


<dt id="arg-force-formula">force_formula<a class="anchor" aria-label="anchor" href="#arg-force-formula"></a></dt>
<dd><p>a logical scalar for whether to bypass model parsimony check.
If <code>force_formula = TRUE</code>, interaction terms whose covariates are not
main effects in the model would be permitted. Results in error if
<code>force_formula = FALSE</code> and <code>length(geno_interact_names) &gt; 0</code>.</p></dd>


<dt id="arg-keep-cellnames">keep_cellnames<a class="anchor" aria-label="anchor" href="#arg-keep-cellnames"></a></dt>
<dd><p>a logical scalar for whether to keep cell barcode names.
If <code>keep_cellnames = TRUE</code>, the memory will be larger. The default is FALSE.</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>a BiocParallelParam class object (from <code>BiocParallel</code> R package)
that must be specified when using <code>parallelization = "biocparallel"</code>. Either
<code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">BiocParallel::SnowParam()</a></code> or <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">BiocParallel::MulticoreParam()</a></code>
can be used to initialize, depending on the operating system. BPPARAM is
not used in other parallelization options. The default is NULL.</p></dd>


<dt id="arg-future-seed">future.seed<a class="anchor" aria-label="anchor" href="#arg-future-seed"></a></dt>
<dd><p>a logical or an integer (of length one or seven), or a list
of length(X) with pre-generated random seeds that can be specified when using
<code>parallelization = "future.apply"</code>. See <code><a href="https://future.apply.futureverse.org/reference/future_lapply.html" class="external-link">future.apply::future_eapply</a></code>
documentation for more details on its usage. future.seed is not used in
other parallelization options. The default is FALSE.</p></dd>


<dt id="arg-data-maxsize">data_maxsize<a class="anchor" aria-label="anchor" href="#arg-data-maxsize"></a></dt>
<dd><p>a positive numeric value used to set max data_list size
in GiB increments. Used only when <code>parallelization = "future.apply"</code>.
The default is 1.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>additional arguments passed to internal functions.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list of named features, each containing a list with the following items:</p><dl><dt><code>fit</code></dt>
<dd><p>a <code>glmmTMB</code> fit object.</p></dd>

<dt><code>time</code></dt>
<dd><p>a numeric scalar for the elapsed time to fit the given feature.</p></dd>

<dt><code>snp_cov</code></dt>
<dd><p>a string scalar or vector of SNP ids used in the
fit for given feature.</p></dd>

<dt><code>model_attr</code></dt>
<dd><p>a list of attributes extracted from each model. NOT
currently implemented.</p></dd>

<dt><code>removed_cell</code></dt>
<dd><p>a string scalar or vector of the cell names
removed due to low-variance (NOT currently implemented).</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>

<div class="section">
<h3 id="parallelization-options">Parallelization options<a class="anchor" aria-label="anchor" href="#parallelization-options"></a></h3>


<p>If "parallel" is used then <code>mclapply</code> is called from the <code>parallel</code> package;
if "biocparallel" is used, then <code>bplapply</code> is called from the <code>BiocParallel</code> package;
if "future.apply" is used, then <code>future_lapply</code> is called from the <code>future.apply</code> package;
if "pbmcapply" is used, then <code>pbmclapply</code> is called from the <code>pbmcapply</code> package.</p>
</div>

    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="cn">NULL</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Y. Dong, Yihui Cen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

